### ISSUE TITLE
Fix BraveSearchTest browser tests in Chromium 92

### ISSUE THREAD

**Issue by mariospr** [brave/brave-browser#16006]
## Description 

**TL;DR**: These tests from [brave_search_browsertest.cc](url) are failing now on DCHECK-enabled builds, because of a [recent change in V8](https://chromium-review.googlesource.com/c/v8/v8/+/2874397) that adds an extra check and, after some initial analysis, it seems that adapting the test would be required.

#### Longer version:

The `BraveSearchTest` browser tests `CheckForAFunction`, `CheckForAFunctionDev` and `CheckForAnUndefinedFunction` from [brave_search_browsertest.cc](https://github.com/brave/brave-core/blob/master/browser/brave_search/brave_search_browsertest.cc) are currently **crashing** on the `cr92` branch **on DCHECK-enabled builds** with the following error:
```
[ RUN      ] BraveSearchTest.CheckForAFunction
[319517:319517:0520/195151.486770:ERROR:power_monitor_device_source_stub.cc(11)] Not implemented reached in virtual bool base::PowerMonitorDeviceSource::IsOnBatteryPower()
[319517:319517:0520/195151.487394:INFO:content_main_runner_impl.cc(1076)] Chrome is running in full browser mode.
[319585:319585:0520/195151.837264:WARNING:vaapi_wrapper.cc(587)] VAAPI video acceleration not available for swiftshader
[319585:319585:0520/195151.846956:ERROR:gpu_init.cc(441)] Passthrough is not supported, GL is swiftshader
[319517:319517:0520/195151.880387:WARNING:external_provider_impl.cc(484)] Malformed extension dictionary for extension: odbfpeeihdkbihmopkbjmoonfanlbfcl.  Key external_update_url has value "", which is not a valid URL.
[319588:319662:0520/195152.209016:WARNING:simple_synchronous_entry.cc(1333)] Could not open platform files for entry.
[319517:319517:0520/195152.214387:ERROR:brave_referrals_service.cc(274)] Failed to fetch headers from referral server, error: -102, response code: -1, payload: , url: https://laptop-updates.brave.com/promo/custom-headers
[319517:319517:0520/195152.391868:INFO:CONSOLE(0)] "Uncaught (in promise) undefined", source: https://search.brave.com:34721/bravesearch.html (0)
#
# Fatal error in ../../v8/src/api/api-inl.h, line 183
# Debug check failed: microtask_queue->GetMicrotasksScopeDepth() || !microtask_queue->DebugMicrotasksScopeDepthIsZero().
#
#
#
#FailureMessage Object: 0x7f60c080db60#0 0x7f60de61ab89 base::debug::CollectStackTrace()
#1 0x7f60de518813 base::debug::StackTrace::StackTrace()
#2 0x7f60d0e245ed gin::(anonymous namespace)::PrintStackTrace()
#3 0x7f60cc6f8e08 V8_Fatal()
#4 0x7f60cc6f8a05 v8::base::(anonymous namespace)::DefaultDcheckHandler()
#5 0x7f60d14bc61e v8::CallDepthScope<>::~CallDepthScope()
#6 0x7f60d14d6fcd v8::Object::Set()
#7 0x56443e09dee1 brave_search::BraveSearchJSHandler::BindFunctionsToObject()
#8 0x56443e09ddcb brave_search::BraveSearchJSHandler::AddJavaScriptObject()
#9 0x56443e09f179 brave_search::BraveSearchServiceWorkerHolder::WillEvaluateServiceWorkerOnWorkerThread()
#10 0x56443fdc152d BraveContentRendererClient::WillEvaluateServiceWorkerOnWorkerThread()
#11 0x7f60dc82d909 content::ServiceWorkerContextClient::WillEvaluateScript()
#12 0x7f60d06590fa blink::ServiceWorkerGlobalScopeProxy::WillEvaluateClassicScript()
#13 0x7f60d58f9965 blink::WorkerGlobalScope::RunWorkerScript()
#14 0x7f60d58f9ab4 blink::WorkerGlobalScope::ReadyToRunWorkerScript()
#15 0x7f60d0646bdd blink::ServiceWorkerGlobalScope::InitializeGlobalScope()
#16 0x7f60d5b72b21 blink::mojom::blink::ServiceWorkerStubDispatch::Accept()
#17 0x7f60dd98eccb mojo::InterfaceEndpointClient::HandleValidatedMessage()
#18 0x7f60dd994eff mojo::MessageDispatcher::Accept()
#19 0x7f60dd990a27 mojo::InterfaceEndpointClient::HandleIncomingMessage()
#20 0x7f60dd99aa62 mojo::internal::MultiplexRouter::ProcessIncomingMessage()
#21 0x7f60dd999fa3 mojo::internal::MultiplexRouter::Accept()
#22 0x7f60dd994eff mojo::MessageDispatcher::Accept()
#23 0x7f60dd987fbd mojo::Connector::DispatchMessage()
#24 0x7f60dd9889f0 mojo::Connector::ReadAllAvailableMessages()
#25 0x7f60dd988799 mojo::Connector::OnWatcherHandleReady()
#26 0x7f60dd989180 mojo::SimpleWatcher::DiscardReadyState()
#27 0x7f60dd93b3ac mojo::SimpleWatcher::OnHandleReady()
#28 0x7f60dd93b8c0 base::internal::Invoker<>::RunOnce()
#29 0x7f60de5ad07f base::TaskAnnotator::RunTask()
#30 0x7f60de5cf441 base::sequence_manager::internal::ThreadControllerWithMessagePumpImpl::DoWorkImpl()
#31 0x7f60de5cebfd base::sequence_manager::internal::ThreadControllerWithMessagePumpImpl::DoWork()
#32 0x7f60de5471e4 base::MessagePumpDefault::Run()
#33 0x7f60de5d000b base::sequence_manager::internal::ThreadControllerWithMessagePumpImpl::Run()
#34 0x7f60de58274a base::RunLoop::Run()
#35 0x7f60d305924d blink::scheduler::WorkerThread::SimpleThreadImpl::Run()
#36 0x7f60de63533e base::(anonymous namespace)::ThreadFunc()
#37 0x7f60cee07609 start_thread
#38 0x7f60ce7dc293 clone
[319517:319517:0520/195152.567710:ERROR:CONSOLE(2562)] "Active tab not found", source: chrome-extension://mnojpmjdmbbfmejpflffifhffcmidifd/out/brave_extension_background.bundle.js (2562)
[319517:319517:0520/195152.583751:WARNING:CONSOLE(18513)] "no tab id so cannot request shield data from window focus change!", source: chrome-extension://mnojpmjdmbbfmejpflffifhffcmidifd/out/brave_extension_background.bundle.js (18513)
[319517:319517:0520/195153.058491:INFO:browser_test_utils.cc(2486)] DOMMessageQueue::RenderProcessGone 3
../../content/public/test/no_renderer_crashes_assertion.cc:101: Failure
Failed

Unexpected termination of a renderer process; status: 3, exit_code: 132
Stack trace:
#0 0x55a3b612cb93 content::NoRendererCrashesAssertion::Observe()
#1 0x7f7d7c44b237 content::NotificationServiceImpl::Notify()
#2 0x7f7d7c61204a content::RenderProcessHostImpl::ProcessDied()
#3 0x7f7d7b26fc0a base::internal::Invoker<>::RunOnce()
#4 0x7f7d7e85207f base::TaskAnnotator::RunTask()
#5 0x7f7d7e874441 base::sequence_manager::internal::ThreadControllerWithMessagePumpImpl::DoWorkImpl()
#6 0x7f7d7e873bfd base::sequence_manager::internal::ThreadControllerWithMessagePumpImpl::DoWork()
#7 0x7f7d7e7ed04a base::MessagePumpGlib::Run()
#8 0x7f7d7e87500b base::sequence_manager::internal::ThreadControllerWithMessagePumpImpl::Run()
#9 0x7f7d7e82774a base::RunLoop::Run()
#10 0x55a3b6118ac3 content::DOMMessageQueue::WaitForMessage()
#11 0x55a3b6116501 content::(anonymous namespace)::ExecuteScriptHelper()
#12 0x55a3b6117156 content::(anonymous namespace)::EvalRunnerScript()
#13 0x55a3b6115d7e content::EvalJs()
#14 0x55a3b334170f BraveSearchTest_CheckForAFunction_Test::RunTestOnMainThread()
#15 0x55a3b6113bf5 content::BrowserTestBase::ProxyRunTestOnMainThreadLoop()
#16 0x55a3b4588bd5 (anonymous namespace)::InvokeClosureAfterGetPluginCallback()
#17 0x55a3b469ce5e ChromeBrowserMainParts::PreMainMessageLoopRunImpl()
#18 0x55a3b469c284 ChromeBrowserMainParts::PreMainMessageLoopRun()
#19 0x7f7d7c164a4f content::BrowserMainLoop::PreMainMessageLoopRun()
#20 0x7f7d7c7405f7 content::StartupTaskRunner::RunAllTasksNow()
#21 0x7f7d7c1646c5 content::BrowserMainLoop::CreateStartupTasks()
#22 0x7f7d7c16753b content::BrowserMainRunnerImpl::Initialize()
#23 0x7f7d7c162f37 content::BrowserMain()
#24 0x7f7d7cc096f2 content::ContentMainRunnerImpl::RunBrowser()
#25 0x7f7d7cc0914d content::ContentMainRunnerImpl::Run()
#26 0x7f7d7cc06d5a content::RunContentProcess()
#27 0x7f7d7cc06e3d content::ContentMain()
#28 0x55a3b611310f content::BrowserTestBase::SetUp()
#29 0x55a3b4fda256 InProcessBrowserTest::SetUp()
[319517:319517:0520/195153.128141:ERROR:browser_test_utils.cc(183)] Cannot communicate with DOMMessageQueue.
../../brave/browser/brave_search/brave_search_browsertest.cc:117: Failure
Expected equality of these values:
  base::Value(true)
    Which is: true
  result_first.value
    Which is: null
Stack trace:
#0 0x55a3b33417d9 BraveSearchTest_CheckForAFunction_Test::RunTestOnMainThread()
#1 0x55a3b6113bf5 content::BrowserTestBase::ProxyRunTestOnMainThreadLoop()
#2 0x55a3b4588bd5 (anonymous namespace)::InvokeClosureAfterGetPluginCallback()
#3 0x55a3b469ce5e ChromeBrowserMainParts::PreMainMessageLoopRunImpl()
#4 0x55a3b469c284 ChromeBrowserMainParts::PreMainMessageLoopRun()
#5 0x7f7d7c164a4f content::BrowserMainLoop::PreMainMessageLoopRun()
#6 0x7f7d7c7405f7 content::StartupTaskRunner::RunAllTasksNow()
#7 0x7f7d7c1646c5 content::BrowserMainLoop::CreateStartupTasks()
#8 0x7f7d7c16753b content::BrowserMainRunnerImpl::Initialize()
#9 0x7f7d7c162f37 content::BrowserMain()
#10 0x7f7d7cc096f2 content::ContentMainRunnerImpl::RunBrowser()
#11 0x7f7d7cc0914d content::ContentMainRunnerImpl::Run()
#12 0x7f7d7cc06d5a content::RunContentProcess()
#13 0x7f7d7cc06e3d content::ContentMain()
#14 0x55a3b611310f content::BrowserTestBase::SetUp()
#15 0x55a3b4fda256 InProcessBrowserTest::SetUp()
[319517:319582:0520/195153.308516:ERROR:process_posix.cc(330)] Unable to terminate process 319588: No such process (3)
[319517:319582:0520/195153.308713:WARNING:internal_linux.cc(66)] Failed to read /proc/319588/stat
[319517:319530:0520/195153.310492:WARNING:discardable_shared_memory_manager.cc(432)] Some MojoDiscardableSharedMemoryManagerImpls are still alive. They will be leaked.
[  FAILED  ] BraveSearchTest.CheckForAFunction, where TypeParam =  and GetParam() =  (2182 ms)
[1/1] BraveSearchTest.CheckForAFunction (2621 ms)
1 test failed:
    BraveSearchTest.CheckForAFunction (../../brave/browser/brave_search/brave_search_browsertest.cc:108)
```

I've been looking into this today and, following the following excerpt from the error above:
```
# Fatal error in ../../v8/src/api/api-inl.h, line 183
# Debug check failed: microtask_queue->GetMicrotasksScopeDepth() || !microtask_queue->DebugMicrotasksScopeDepthIsZero().
```
... I found that the problem was probably already there before Cr92, and only got unveiled now after following change in V8:
[https://chromium-review.googlesource.com/c/v8/v8/+/2874397](https://chromium-review.googlesource.com/c/v8/v8/+/2874397)

FWIW, the `DCHECK` that is now failing (and therefore crashing the renderer) was not reached before because it was guarded by `V8_CHECK_MICROTASKS_SCOPES_CONSISTENCY` , which was always false for us. But it's now guarded by `DEBUG` and thus we're reaching it, leading to the crash.

I've talked to one V8 dev and tried a few things to see if there could be some place in `browser_test_utils.cc `(or elsewhere) missing something like setting `v8::MicrotasksScope::kDoNotRunMicrotasks`, but that didn't seem to work.

Then I've been carefully looking into the code in `brave/browser/brave_search/brave_search_browsertest.cc`, as well as in the `brave_search.html` and `sw.js` files, and it seems like there might be something fundamentally wrong with the test, which does register the SW (at least apparently) but then does not like installing the event listener as part of a `EvalJS()` call.

The (in principle) good news is that talking to it seems that all we might need to could be to adapt the test, meaning that this could be a test-only (and not production-related) issue.

Since most changes related to the [JS side of those tests were made just a bit over a month ago](https://github.com/brave/brave-core/commits/master/browser/brave_search/brave_search_browsertest.cc?author=SergeyZhukovsky) by @SergeyZhukovsky, I've talked to him and I'm **disabling the tests for now + assigning this issue to him**, as agreed over Slack.

**Comment by SergeyZhukovsky** [comment]
The fix is here https://github.com/brave/brave-core/commit/27aaf9ef6f9622c004f8defb60a74d6d400c8572
